<a href="https://colab.research.google.com/github/Yagut21/Data-Science/blob/main/Homework4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/ogut77/DataScience/master/insurance.csv"
df = pd.read_csv(url)



In [2]:
df.head()


age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

Context in Insurance Data
This dataset is often used to predict charges based on the other variables (age, sex, bmi, children, smoker, region). For example:

Input Variables (X): age, sex, bmi, children, smoker, region (features used to make predictions).

Output Variable (y): charges (what you’re trying to predict).

Describtion of variables
1. Age
Description: The age of the individual (the insured person).
Type: Numerical (integer).
Example Values: 19, 45, 62, etc.
Role in Insurance: Age is a key factor in determining insurance charges. Older individuals often have higher medical costs (and thus higher charges) due to increased health risks.
2. Sex
Description: The gender of the individual.
Type: Categorical (text or binary).
Example Values: "male," "female"
Role in Insurance: Gender can influence insurance charges because health risks and medical expenses may differ between males and females (e.g., pregnancy-related costs for females).
3. BMI (Body Mass Index)
Description: A measure of body fat based on height and weight (calculated as weight in kg divided by height in meters squared).
Type: Numerical (float).
Example Values: 25.3, 30.1, 18.5, etc.
Role in Insurance: Higher BMI often correlates with increased health risks (e.g., obesity-related conditions like diabetes or heart disease), leading to higher insurance charges.
4. Children
Description: The number of children (dependents) covered under the individual’s insurance plan.
Type: Numerical (integer).
Example Values: 0, 1, 3, etc.
Role in Insurance: More children can increase insurance costs slightly, as it may reflect additional healthcare needs, though the effect is often less pronounced than other factors like smoking or age.
5. Smoker
Description: Indicates whether the individual smokes tobacco.
Type: Categorical (text or binary).
Example Values: "yes," "no" .
Role in Insurance: Smoking is a major factor in insurance charges. Smokers typically have much higher medical costs due to risks like lung disease or cancer, so their charges are significantly elevated.
6. Region
Description: The geographic region where the individual lives.
Type: Categorical (text).
Example Values: "northeast," "southeast," "southwest," "northwest" (common in U.S.-based datasets).
Role in Insurance: Charges can vary by region due to differences in healthcare costs, lifestyle factors, or local insurance regulations.
7. Charges
Description: The insurance charges (or premiums/costs) billed to the individual, typically in a currency like USD.
Type: Numerical (float).
Example Values: 1684.52, 11234.89, 32050.23, etc.
Role in Insurance: This is usually the target variable (output) in predictive modeling. It represents the amount the insurance company charges, influenced by all the other columns (age, sex, BMI, etc.).



In [3]:
#1. Check if there is null value in dataset df (5 pt)
df.isnull().sum()


age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [4]:
#2. Assign charges to y  and others to X using df. y is output variable and X is input variables (5 pt)
y = df["charges"]
X = df.drop("charges", axis=1)

In [7]:
#3. Use  get_dummies() function from the pandas library to convert categorical variables in a DataFrame (X).
# Drop first drops the first category’s dummy variable to avoid multicollinearity (5 pt)
X = pd.get_dummies(X, drop_first=True)

In [8]:
#Use following methods for the evaluation on test and train data
def evalmetric(y,ypred):
 from scipy.stats import pearsonr
 import numpy as np
 e = y - ypred
 mse_f = np.mean(e**2)
 rmse_f = np.sqrt(mse_f)
 mae_f = np.mean(abs(e))
 mape_f = 100*np.mean(abs(e/y))
 crl, _ = pearsonr(y, ypred)
 r2_f = crl*crl
 print("MSE:", mse_f)
 print("RMSE:", rmse_f)
 print("MAE:",mae_f)
 print("MAPE:",mape_f)
 print("R-Squared:", round(r2_f, 4))


In [9]:
#4.Get the correlation between X variables and y variables.(5 pt)
correlations = X.corrwith(y)
print(correlations)

age                 0.299008
bmi                 0.198341
children            0.067998
sex_male            0.057292
smoker_yes          0.787251
region_northwest   -0.039905
region_southeast    0.073982
region_southwest   -0.043210
dtype: float64


In [10]:
#5.Split a dataset into 25%  of data as test data  and 75% of data as training data ( pt)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [11]:
#6. Using Decision Tree and Linear Regression methods, compare the performance results on both test and training data
#to determine which one is more likely to overfit and which is more likely to underfit.
# Do you think that Lasso and Ridge regularization are more likely to improve the results of Linear model test data,
# or would Random Forest or Boosting methods are more likely to improve the results of Decison tree test data?
#Explain your reasoning.(35 pt)
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error
dt_model = DecisionTreeRegressor(random_state=42)
dt_model.fit(X_train, y_train)

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

y_train_pred_dt = dt_model.predict(X_train)
y_test_pred_dt = dt_model.predict(X_test)

y_train_pred_lr = lr_model.predict(X_train)
y_test_pred_lr = lr_model.predict(X_test)

print("Decision Tree Performance:")
print("---Train:")
evalmetric(y_train, y_train_pred_dt)
print("---Test:")
evalmetric(y_test, y_test_pred_dt)

print("\nLinear Regression Performance:")
print("---Train:")
evalmetric(y_train, y_train_pred_lr)
print("---Test:")
evalmetric(y_test, y_test_pred_lr)

Decision Tree Performance:
---Train:
MSE: 182648.17106092346
RMSE: 427.37357318969015
MAE: 19.084122482552342
MAPE: 0.47775094523744194
R-Squared: 0.9987
---Test:
MSE: 36580230.16400095
RMSE: 6048.159237652473
MAE: 2693.294304847761
MAPE: 29.869257985320623
R-Squared: 0.7805

Linear Regression Performance:
---Train:
MSE: 37004502.18409475
RMSE: 6083.132596294014
MAE: 4183.153367011969
MAPE: 42.26867490005659
R-Squared: 0.745
---Test:
MSE: 35117755.73613632
RMSE: 5926.023602394469
MAE: 4243.654116653137
MAPE: 44.468185116980976
R-Squared: 0.7676


1. An extensive gap between training set performance and testing set performance confirms that overfitting occurred.
2. Low bias problems in the linear regression model create minimal impact on the error reduction potential and feature selection process of Lasso and Ridge regularization approaches. Test errors of the decision tree model demonstrate obvious overfitting through its low training errors and superior test errors. Random Forest and Boosting methods demonstrate potential for better test performance through their ability to use multiple trees which help control data variance and detect complex non-linear patterns.

In [13]:
#7. Explain performance of linear regressin on test data
# using  Root mean squared error, mean absolute error, mean absolute percentage error and R2 metric (10 pt)

The linear regression model delivers a moderate performance result in test evaluation where it accounts for 76.76% of the variance based on R². The predictive model exhibits major deviation based on the high RMSE (5926.02) and MAE (4243.65) values which indicate significant differences between model predictions and actual values. Moreover the MAPE (44.47%) illustrates substantial percentage error. The model displays essential data patterns but its prediction accuracy requires improvement because it does not meet optimal standards. But if decision tree method was used, the model could be better.

In [12]:
#8. Use Random Forest and Boosting methods (XGBoost, LightGBM, and CatBoost)
#to obtain the evaluation scores on  test data.
#Which Boosting technique yielded the best performance on the test data based on the R² metric?
#Did you achieve a better result compared to Random Forest on the test data based on the R² metric?
#If there is improvement on Random forest or boosting methods over decison tree, explain  (30 pt)
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
!pip install catboost==1.2.7
from catboost import CatBoostRegressor

rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

xgb_model = XGBRegressor(random_state=42, objective='reg:squarederror')
xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)

lgb_model = LGBMRegressor(random_state=42)
lgb_model.fit(X_train, y_train)
y_pred_lgb = lgb_model.predict(X_test)

cat_model = CatBoostRegressor(random_state=42, verbose=0)
cat_model.fit(X_train, y_train)
y_pred_cat = cat_model.predict(X_test)

print("\nTest Performances:")

print("\nRandom Forest:")
evalmetric(y_test, y_pred_rf)

print("\nXGBoost:")
evalmetric(y_test, y_pred_xgb)

print("\nLightGBM:")
evalmetric(y_test, y_pred_lgb)

print("\nCatBoost:")
evalmetric(y_test, y_pred_cat)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 MB 7.0 MB/s eta 0:00:00
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001170 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 319
[LightGBM] [Info] Number of data points in the train set: 1003, number of used features: 8
[LightGBM] [Info] Start training from score 13267.935814

Test Performances:

Random Forest:
MSE: 23114410.14038345
RMSE: 4807.7448081593775
MAE: 2653.61478095801
MAPE: 30.27386314379988
R-Squared: 0.8506

XGBoost:
MSE: 26433443.13176504
RMSE: 5141.346431798293
MAE: 2957.213261792119
MAPE: 34.57266690344694
R-Squared: 0.8301

LightGBM:
MSE: 22005117.949021608
RMSE: 4690.961303296117
MAE: 2700.720352413941
MAPE: 33.101463593940274
R-Squared: 0.8553

CatBoost:
MSE: 21340838.308785602
RMSE: 4619.614519501124
MAE: 2608.117710576109
MAPE: 30.82075085150296
R-Squared: 0.8

CatBoost demonstrated superior performance compared to LightGBM and XGBoost since it produced an R² of 0.8589 on test data whilst LightGBM achieved 0.8553 and XGBoost reached 0.8301. The test data results indicated enhancements compared to Random Forest because it reached an R² of 0.8506. The performance enhancement achieved by boosting methods occurs through sequential error correction between different models while the Random Forest method uses bagging to obtain averaged decision trees. The alternative ensemble approaches combining Random Forest model with boosting demonstrate much better performance when compared to the single decision tree model that exhibited a test R² of 0.7805 due to overfitting complications.